vllm/v1/executor/multiproc_executor.py

```diff
@@ -0,0 +1,171 @@
+            f"tensor_parallel_size ({tensor_parallel_size}). "
+            f"Pipeline parallelism is not yet implemented in v1")
+
+        # Set multiprocessing envs that are common to V0 and V1
+        set_multiprocessing_worker_envs(self.parallel_config)
+
+        # Multiprocessing-based executor does not support multi-node setting.
+        # Since it only works for single node, we can use the loopback address
+        # 127.0.0.1 for communication.
+        distributed_init_method = get_distributed_init_method(
+            "127.0.0.1", get_open_port())
+
+        # Initialize worker and set up message queues for SchedulerOutputs
+        # and ModelRunnerOutputs
+        self.worker_request_mq = MessageQueue(world_size, world_size)
+        scheduler_output_handle = self.worker_request_mq.export_handle()
+
+        # Create workers
+        self.workers: List[WorkerProcHandle] = []
+        for rank in range(world_size):
+            worker = WorkerProc.make_worker_process(vllm_config, rank, rank,
+                                                    distributed_init_method,
+                                                    scheduler_output_handle)
+            self.workers.append(worker)
+
+    def initialize(self, num_gpu_blocks: int) -> None:
+        """
+        Initialize the KV caches and begin the model execution loop of the
+        underlying workers.
+        """
+        with ThreadPoolExecutor() as executor:
+            futures = [
+                executor.submit(w.initialize, num_gpu_blocks)
+                for w in self.workers
+            ]
+            success_vals = [f.result()
+                            for f in futures]  # Wait for all to complete
+
+        if not all(success_vals):
+            raise RuntimeError("Worker initialization failed.")
+
+        self.worker_request_mq.wait_until_ready()
+        for w in self.workers:
+            w.worker_response_mq.wait_until_ready()
+
+    def determine_num_available_blocks(self) -> Tuple[int, int]:
+        """
+        Determine the number of available KV blocks by invoking the
+        underlying worker.
+        """
+        # Get the maximum number of blocks that can be allocated on GPU and CPU.
+        with ThreadPoolExecutor() as executor:
+            futures = [
+                executor.submit(w.determine_num_available_blocks)
+                for w in self.workers
+            ]
+            num_blocks = [f.result()
+                          for f in futures]  # Wait for all to complete
+
+        # Since we use a shared centralized controller, we take the minimum
+        # number of blocks across all workers to make sure all the memory
+        # operators can be applied to all workers.
+        num_gpu_blocks = min(b[0] for b in num_blocks)
+        num_cpu_blocks = min(b[1] for b in num_blocks)
+
+        return num_gpu_blocks, num_cpu_blocks
+
+    def collective_rpc(self, fn: str, timeout: Optional[float], *args,
```

Comment:
we should also add `output_ranks` option, it's not always the first rank returns something, users can request the output from multiple workers.

and since we have `responses`, it should be easy to just return corresponding slices.